src/Engine/Util/Memory.hpp

```diff
@@ -0,0 +1,240 @@
+#pragma once
+
+
+
+#define USE_RPMALLOC
+
+
+
+#include <scoped_allocator>
+#include <string>
+#include <vector>
+#include <list>
+#include <set>
+#include <unordered_set>
+#include <map>
+#include <unordered_map>
+
+
+
+#ifdef USE_RPMALLOC
+
+
+
+#include "ThirdParty/CoherentLabs_rpmalloc/rpmalloc.h"
+
+
+
+// trick to call memory initialization functions before static evaluations
+namespace detail {
+
+struct InitializeMemory {
+
+    InitializeMemory() {
+        if (!initialized++) {
+            coherent_rpmalloc::rpmalloc_initialize();
+        }
+    }
+
+    static int initialized;
+
+};
+
+static InitializeMemory f_initializeMemory;
+
+}
+
+
+
+#endif
+
+
+
+inline void * allocate(size_t size) {
+#ifdef USE_RPMALLOC
+    return coherent_rpmalloc::rpmalloc(size);
+#else
+    return std::malloc(size);
+#endif
+}
+
+inline void deallocate(void * ptr) {
+#ifdef USE_RPMALLOC
+    return coherent_rpmalloc::rpfree(ptr);
+#else
+    return std::free(ptr);
+#endif
+}
+
+
+
+template <typename T>
+struct Allocator {
+
+    template <typename U> friend struct Allocator;
+
+    using value_type = T;
+    using pointer = T *;
+
+    using propagate_on_container_copy_assignment = std::true_type;
+    using propagate_on_container_move_assignment = std::true_type;
+    using propagate_on_container_swap = std::true_type;
+
+    Allocator() = default;
+
+    ~Allocator() = default;
+
+    template <typename U> Allocator(const Allocator<U> &) {}
+
+    pointer allocate(std::size_t n) {
+        return static_cast<pointer>(::allocate(n * sizeof(T)));
+    }
+
+    void deallocate(pointer p, std::size_t n) {
+        ::deallocate(p);
+    }
+
+};
+
+template <typename T1, typename T2>
+bool operator==(const Allocator<T1> & a1, const Allocator<T2> & a2) {
+    return true;
+}
+
+template <typename T1, typename T2>
+bool operator!=(const Allocator<T1> & a1, const Allocator<T2> & a2) {
+    return false;
+}
+
+template <typename T> using ScopedAllocator = std::scoped_allocator_adaptor<Allocator<T>>;
+
+
+
+// Standard containers using custom memory allocator
+
+// std::string equivalent
+using String = std::basic_string<char, std::char_traits<char>, ScopedAllocator<char>>;
+
+// convert String to std::string
+inline std::string convert(const String & string) {
+    return std::string(string.c_str());
+}
+
+// convert std::string to String
+inline String convert(const std::string & string) {
+    return String(string.c_str());
+}
+
+// std::vector equivalent
```

Comment:
Are all of these necessary for rpmalloc? Seems confusing and unnecessary otherwise.

Reply:
the standard library containers that dynamically allocate memory all have an allocator template type parameter that defaults to std::allocator<T>. This is what you replace if you want them to use a different allocator. These are just helper typdefs so instead of typing std::vector<T, ScopedAllocator<T>> you can just type Vector<T>